docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,585 @@
+# C++ code style guidelines
+
+We use C++14, but currently consider C++17 too immature for usage.
+
+This documents contains guidelines regarding:
+
+* File structure
+* Comments and documentation
+* Tests
+* Naming
+* Formatting
+* Typing
+* Namespaces
+* Structs
+* Classes
+* Error handling
+* Memory handling
+* Resource handling
+* Other
+
+
+## File structure
+
+* Header files: `.hpp`
+* Source files: `.cpp`
+* Use `#pragma once` in headers.
+* Split class definition across a header and source file by default.
+* Include header definitions in the same file.
+
+* Order of includes
+	1. Related header.
+	2. `C` library.
+	3. `C++` library.
+	4. Other libraries.
+	5. Fetch library.
+
+* Classes in this order
+	1. `public`.
+	2. `protected`.
+	3. `private`.
+
+
+## Comments and documentation
+
+* Provide API documentation with your code.
+* Include a comment summary at the top of all files.
+
+``` c++
+/* Approximation of the exponential function.
```

Comment:
I think that this style is a little inconsistent with the current project. I think maybe we find a better example

Reply:
After the indentation changes etc, does this still look inconsistent?